### PR TITLE
KEYCLOAK-14023 Update InstagramIdentityProvider.java

### PR DESCRIPTION
**Pull request by jbpin**
Update Instagram provider to support instagram graph api. Unfortunately we can't retrieve user's name.

### CODE REVIEW

**Comment by stianst** [comment]
Please explain why you are proposing this change

**Comment by jbpin** [comment]
Hi @stianst. Instagram user endpoint have change: 
https://developers.facebook.com/docs/instagram-basic-display-api/reference/me

Actually I'm not sure we should change it or create a new one ?

**Comment by jbpin** [comment]
Maybe related to this: https://issues.redhat.com/browse/KEYCLOAK-7761

**Comment by jbpin** [comment]
I've created a new Issue: https://issues.redhat.com/browse/KEYCLOAK-14023

**Comment by hguerrero** [comment]
@stianst can we merge this? it should be fixing the problem with the instagram login as described correctly in the JIRA

File: services/src/main/java/org/keycloak/social/instagram/InstagramIdentityProvider.java [hunk 1]
```diff
@@ -46,23 +47,20 @@ public InstagramIdentityProvider(KeycloakSession session, OAuth2IdentityProvider
...
+				.param("access_token", accessToken)
+				.param("fields", PROFILE_FIELDS)
+				.asJson();
 			
 			JsonNode profile = raw.get("data");
 			
 			logger.debug(profile.toString());
 
 			String id = getJsonProperty(profile, "id");
+	  		String username = getJsonProperty(profile, "username");
```
Comment (JavaSwitch) [inline services/src/main/java/org/keycloak/social/instagram/InstagramIdentityProvider.java:60]: raw no 'data' of node

```suggestion
			JsonNode raw = SimpleHttp.doGet(PROFILE_URL,session)
				.param("access_token", accessToken)
				.param("fields", PROFILE_FIELDS)
				.asJson();
			
			logger.debug(raw.toString());

			String id = getJsonProperty(raw, "id");
	  		String username = getJsonProperty(raw, "username");
```

**Review by huangwenkan9 — APPROVED** [review]

File: services/src/main/java/org/keycloak/social/instagram/InstagramIdentityProvider.java (hunk 1 above)
Comment (JavaSwitch) [inline services/src/main/java/org/keycloak/social/instagram/InstagramIdentityProvider.java:60]: raw not 'data' of node
```suggestion
			logger.debug(raw.toString());
			
			String id = getJsonProperty(raw, "id");
	  		String username = getJsonProperty(raw, "username");
```

**Comment by vmuzikar** [comment]
Will try to review by the end of this week.

**Review by vmuzikar — COMMENTED** [review]
@jbpin Thank you for your contribution. What's the motivation to make this change? The Instagram IdP seems to be working as expected and the [used API](https://www.instagram.com/developer/endpoints/users/#get_users_self) doesn't seem to be deprecated.

**Comment by vmuzikar** [comment]
Scratch that. The old API is really deprecated and will get disabled in a few days(!)

**Comment by hguerrero** [comment]
So yes, the old API is deprecated and moved to the new Facebook Graph for most of the old features. However there is a basic API that includes the OAuth flow to autenticate. This API won't return the name but still returns an username at works for the purpose of brokering.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@jbpin Thank you for your contribution.

As @JavaSwitch already mentioned, this requires fixing which I can confirm.

Also, we need to address migration. Let's say someone uses older Keycloak version that uses legacy Instagram API. Then upgrades to newer Keycloak version to use the new Instagram API but users are now unable to login using Instagram, because the new API uses different user IDs. Luckily, per the documentation [we can get the old legacy ID](https://developers.facebook.com/docs/instagram-basic-display-api). Those legacy IDs will be however available only temporarily, so the IdP in Keycloak will have to update it to the new ID.

This will also require updating the [documentation](https://www.keycloak.org/docs/latest/server_admin/#social-identity-providers) (incl. the note about legacy IDs).

As this is quite critical fix, please let us know if you're able to work on this.

**Comment by vmuzikar** [comment]
Since this is quite an urgent matter, we'll take changes from this PR and finish them up in a new PR.

**Comment by jbpin** [comment]
> Since this is quite an urgent matter, we'll take changes from this PR and finish them up in a new PR.

Sorry I don't have time to work on this one.

**Comment by vmuzikar** [comment]
@jbpin No worries! :) Thanks again for your contribution, I'll use it in the new PR.

**Comment by abstractj** [comment]
Superseded by #7219

**Pull request closed without merging**